File: engine/rules/saves.py
"""Saving throw resolution."""

from __future__ import annotations

from dataclasses import dataclass

from engine.core.rng import DeterministicRNG
from engine.rules.checks import CheckResult, resolve_check
from engine.rules.degrees import Degree


@dataclass
class SaveProfile:
    fortitude: int = 0
    reflex: int = 0
    will: int = 0


def resolve_save(rng: DeterministicRNG, save_type: str, profile: SaveProfile, dc: int) -> CheckResult:
    lookup = {
        "Fortitude": profile.fortitude,
        "Reflex": profile.reflex,
        "Will": profile.will,
    }
    modifier = lookup.get(save_type, 0)
    return resolve_check(rng=rng, modifier=modifier, dc=dc)


def basic_save_multiplier(degree: Degree) -> float:
    if degree == "critical_success":
        return 0.0
    if degree == "success":
        return 0.5
    if degree == "failure":
        return 1.0
    return 2.0

File: src/rules/saves.ts
/**
 * Saving throw resolution.
 */

import { DeterministicRNG } from "../engine/rng";
import { CheckResult, resolveCheck } from "./checks";
import { Degree } from "./degrees";

export interface SaveProfile {
  fortitude: number;
  reflex: number;
  will: number;
}

export function resolveSave(
  rng: DeterministicRNG,
  saveType: string,
  profile: SaveProfile,
  dc: number,
): CheckResult {
  const normalized = saveType.charAt(0).toUpperCase() + saveType.slice(1).toLowerCase();
  const lookup: Record<string, number> = {
    Fortitude: profile.fortitude,
    Reflex: profile.reflex,
    Will: profile.will,
  };
  const modifier = lookup[normalized] ?? 0;
  return resolveCheck(rng, modifier, dc);
}

export function basicSaveMultiplier(degree: Degree): number {
  if (degree === "critical_success") return 0.0;
  if (degree === "success") return 0.5;
  if (degree === "failure") return 1.0;
  return 2.0; // critical_failure
}
